addon/adapters/debug-adapter.js

```diff
@@ -0,0 +1,269 @@
+import { A } from '@ember/array';
+import { inject } from '@ember/service';
+import M3TrackedArray from '../m3-tracked-array';
+import M3ReferenceArray from '../m3-reference-array';
+
+/*
+  Extend `Ember.DataAdapter` with m3 specific code.
+
+  @class DebugAdapter
+  @extends Ember.DataAdapter
+  @private
+*/
+
+// TODO: implement getFilters/getRecordColor/getRecordFilterValues (for model state in the cache)
+// and getRecordKeywords (for search)
+export default DataAdapter.extend({
+  schema: inject('m3-schema'),
+
+  init() {
+    this._super(...arguments);
+    // This keeps track of all model types the debug adapter has seen already (so we don't watch for changes twice)
+    this.seenTypes = {};
+  },
+
+  /**
+    Get the values from a record and converts them as needed to be displayed as column values
+    Depending on the type of data it may need to be transformed into JSON or can just be returned as a string
+    @private
+    @method _getRecordValues
+    @param attributeValue nested record data that needs to be processed
+    @return data to be rendered in the column values
+  */
+  _getRecordValues(attributeValue) {
+    const attributeValueIsArray =
+      attributeValue instanceof M3TrackedArray || attributeValue instanceof M3ReferenceArray;
+
+    if (attributeValue instanceof MegamorphicModel) {
+      return attributeValue.debugJSON();
+    } else if (attributeValueIsArray && get(attributeValue, 'length')) {
+      return attributeValue.map(nestedAttributeValue =>
+        this._getRecordValues(nestedAttributeValue)
+      );
+    } else if (attributeValueIsArray && !get(attributeValue, 'length')) {
+      return [];
+    } else if (attributeValue instanceof Object) {
+      return JSON.parse(JSON.stringify(attributeValue));
+    }
+    return attributeValue;
+  },
+
+  _nameToClass(modelName) {
+    return modelName;
+  },
+
+  /**
+    Fetches all Megamorphic model types
+    @private
+    @method getModelTypes
+    @return {Array} Array of model types
+  */
+  getModelTypes() {
+    let allModelNames = Object.keys(this.store._identityMap._map);
+    let modelNames = allModelNames.filter(
+      modelName => this.store._modelFactoryFor(modelName).class === MegamorphicModel
+    );
+
+    const modelTypes = modelNames.map(name => {
+      // we need to keep klass even though it is not technically needed/correct
+      // because both Ember inspector and Ember's generic debug adapter expect this data structure
+      // and use klass to generate objectIds and such
+      return { klass: name, name };
+    });
+    return modelTypes;
+  },
+
+  /**
+    Transforms a model's attribute name to be a human readable string
+    @private
+    @method columnNameToDesc
+    @param {String} name attribute name from a model type
+    @return {String} Human readable attribute name
+  */
+  columnNameToDesc(name) {
+    return capitalize(
+      underscore(name)
+        .replace(/_/g, ' ')
+        .trim()
+    );
+  },
```

Comment:
This is consistent with how they render the column names in `Ember Data` but I agree it may get confusing/doesn't provide that much value. I guess the question is it better to be accurate vs have the same experience if someone in the future is toggling between `Ember Data` vs `m3` records?

Reply:
I think I'd rather we do the more accurate thing